source/common/rds/basic/route_config_provider_manager_impl.h

```diff
@@ -0,0 +1,87 @@
+          class NullConfigImpl>
+class RouteConfigProviderManagerImpl : public RouteConfigProviderManager<Rds, RouteConfiguration>,
+                                       public Singleton::Instance {
+public:
+  RouteConfigProviderManagerImpl(Server::Admin& admin)
+      : manager_(admin, downCase(getRdsName()) + "_routes", proto_traits_) {}
+
+  // RouteConfigProviderManager
+  RouteConfigProviderSharedPtr createRdsRouteConfigProvider(
+      const Rds& rds, Server::Configuration::ServerFactoryContext& factory_context,
+      const std::string& stat_prefix, Init::Manager& init_manager) override {
+    return manager_.addDynamicProvider(
+        rds, rds.route_config_name(), init_manager,
+        [&factory_context, &rds, &stat_prefix, this](uint64_t manager_identifier) {
+          auto config_update = std::make_unique<RouteConfigUpdateReceiverImpl>(
+              config_traits_, proto_traits_, factory_context);
+          auto resource_decoder =
+              std::make_unique<Envoy::Config::OpaqueResourceDecoderImpl<RouteConfiguration>>(
+                  factory_context.messageValidationContext().dynamicValidationVisitor(),
+                  getNameFieldName());
+          auto subscription = std::make_shared<RdsRouteConfigSubscription>(
+              std::move(config_update), std::move(resource_decoder), rds.config_source(),
+              rds.route_config_name(), manager_identifier, factory_context,
+              stat_prefix + downCase(getRdsName()) + ".", upCase(getRdsName()), manager_);
+          auto provider = std::make_shared<RdsRouteConfigProviderImpl>(std::move(subscription),
+                                                                       factory_context);
+          return std::make_pair(provider, &provider->subscription().initTarget());
+        });
+  }
+
+  RouteConfigProviderPtr createStaticRouteConfigProvider(
+      const RouteConfiguration& route_config,
+      Server::Configuration::ServerFactoryContext& factory_context) override {
+    return manager_.addStaticProvider([&factory_context, &route_config, this]() {
+      return std::make_unique<StaticRouteConfigProviderImpl>(route_config, config_traits_,
+                                                             factory_context, manager_);
+    });
+  }
+
+private:
+  Envoy::Rds::RouteConfigProviderManager manager_;
+  ConfigTraitsImpl<RouteConfiguration, ConfigImpl, NullConfigImpl> config_traits_;
+  ProtoTraitsImpl<RouteConfiguration, 1> proto_traits_;
+
+  std::string getRdsName() { return Rds().GetDescriptor()->name(); }
+
+  std::string getNameFieldName() {
+    ASSERT(RouteConfiguration().GetDescriptor()->FindFieldByNumber(NameFieldNumber));
+    return RouteConfiguration().GetDescriptor()->FindFieldByNumber(NameFieldNumber)->name();
+  }
+
+  std::string downCase(std::string s) {
+    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::tolower(c); });
```

Comment:
This seems a bit unnecessary and specific, can we just have two strings rather than trying to transform them?

Reply:
Unfortunately it is not possible to pass string literal as template argument. And I did not want to get the protocol specific  strings required for `Rds::RdsRouteConfigSubscription` or `Rds::RouteConfigProviderManager` through the constructor of the template (or similar workaround).
In this way at template specialization we get a ready to use class.
The rds proto message name is mixed case. I just thought `somefilter.trds.someroute.config_reload` and `"Unexpected TRDS configuration ...."` looks better than `somefilter.Trds.someroute.config_reload` and `"Unexpected Trds configuration ...."`.

Reply:
OK, but can you switch to the Abseil variants then, e.g. `AsciiStrToLower`?

Reply:
Done.